group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java

```diff
@@ -234,67 +257,93 @@ private GroupAssignment newAssignmentHeterogeneous(
      */
     // Visible for testing
     void roundRobinAssignment(
         Collection<String> memberIds,
-        List<TopicIdPartition> unassignedPartitions,
+        List<TopicIdPartition> partitionsToAssign,
         Map<TopicIdPartition, List<String>> assignment
     ) {
         // We iterate through the target partitions and assign a memberId to them. In case we run out of members (members < targetPartitions),
         // we again start from the starting index of memberIds.
         Iterator<String> memberIdIterator = memberIds.iterator();
-        for (TopicIdPartition targetPartition : unassignedPartitions) {
+        for (TopicIdPartition topicPartition : partitionsToAssign) {
             if (!memberIdIterator.hasNext()) {
                 memberIdIterator = memberIds.iterator();
             }
             String memberId = memberIdIterator.next();
-            assignment.computeIfAbsent(targetPartition, k -> new ArrayList<>()).add(memberId);
+            assignment.computeIfAbsent(topicPartition, k -> new ArrayList<>()).add(memberId);
+        }
+    }
+
+    /**
+     * This functions assigns topic partitions to members by a round-robin approach and updates the existing assignment.
+     * @param memberIds              The member ids to which the topic partitions need to be assigned, should be non-empty.
+     * @param partitionsToAssign     The subscribed topic partitions which needs assignment.
+     * @param assignment             The existing assignment by topic partition. We need to pass it as a parameter because this
+     *                               method can be called multiple times for heterogeneous assignment.
+     * @param desiredAssignmentCount The number of partitions which can be assigned to each member to give even balance.
+     *                               Note that this number is slightly higher than strictly required to allow for situations
+     *                               in which we have hashing collisions.
+     */
+    void roundRobinAssignmentWithCount(
+        Collection<String> memberIds,
+        List<TopicIdPartition> partitionsToAssign,
+        Map<String, Set<TopicIdPartition>> assignment,
+        int desiredAssignmentCount
+    ) {
+        Collection<String> memberIdsCopy = new LinkedHashSet<>(memberIds);
+
+        // We iterate through the target partitions which are not in the assignment and assign a memberId to them.
+        // In case we run out of members (memberIds < partitionsToAssign), we again start from the starting index of memberIds.
+        Iterator<String> memberIdIterator = memberIdsCopy.iterator();
+        ListIterator<TopicIdPartition> partitionListIterator = partitionsToAssign.listIterator();
+        while (partitionListIterator.hasNext()) {
+            TopicIdPartition partition = partitionListIterator.next();
+            if (!memberIdIterator.hasNext()) {
+                memberIdIterator = memberIdsCopy.iterator();
```

Comment:
I see we have this cycling iterator pattern twice already. I imagine it's a common pattern for things like assignment. Maybe we should create a utility method/class that provides this functionality and adds things like the infinite loop check below. Doesn't have to be done in this PR, just an idea